### PR TITLE
Make DeGraph check faster and more user friendly

### PR DESCRIPTION
**Pull request by oehme**
## Overview

This is a first step towards a Degraph Gradle plugin as discussed in #483. It adds a `degraph` task to every project which ensures that the code is cycle-free. We no longer need to scan the whole classpath as each project is checked individually. This makes builds quite a bit faster.

As you can see the DegraphCheck task is still hardcoded with the configuration for the JUnit project. We'd need to discuss how to model these constraints. In the simplest case we could just have a dedicated configuration file, like Degraph [already has](http://blog.schauderhaft.de/degraph/documentation.html#configuration-file) for visulization.

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

### CODE REVIEW

File: gradle/degraph.gradle [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+			.noJars()
+			.including("org.junit.platform.**")
+			.including("org.junit.vintage.**")
+			.including("org.junit.jupiter.**")
+			.withSlicing("module",
+				new NamedPattern("org.junit.vintage.engine.**", "junit-vintage-engine"),
+				new NamedPattern("org.junit.jupiter.api.**", "junit-jupiter-api"),
+				new NamedPattern("org.junit.jupiter.engine.**", "junit-jupiter-engine"),
+				"org.junit.platform.(*).**"
+			)
```
Comment (marcphilipp) [inline gradle/degraph.gradle]: Does definining the slicing still make sense?

Comment (oehme) [inline gradle/degraph.gradle]: As far as I can see it is there to provide nicer names in error messages/graphs.

Comment (marcphilipp) [inline gradle/degraph.gradle]: I'm fine with leaving it in there for now.

**Comment by oehme** [comment]
> Thanks, @oehme! Have you measured how much this speeds up the build?

`gradle clean degraph` is down from 16s to 2.5s. So total build time is reduced by 13.5s.

> In theory, there's a difference between checking each subproject separately and everything at once. Multiple subprojects could contain classes in the same package and thereby introduce a package cycle that is not present when checking them individually. 

I'd consider split package a really bad practice in itself, but I'm sure we could make the plugin flexible enough to check for that too.

**Comment by marcphilipp** [comment]
> I'd consider split package a really bad practice in itself.

Agreed!

**Comment by marcphilipp** [comment]
Thanks again, @oehme! 👍

**Comment by sbrannen** [comment]
Thanks, @oehme!

**Comment by bechte** [comment]
Cool. Thank you. @oehme

**Pull request merged**